logger/analytics/analyticslogger.go

```diff
@@ -89,22 +115,66 @@ func (al *Logger) Write(bs []byte) (int, error) {
 		return 0, err
 	}
 	bs = append(bs, '\n')
-	r := retrier.New(retrier.ExponentialBackoff(5, 100*time.Millisecond), RequestErrorClassifier{})
-	if err := r.Run(func() error {
-		_, err := al.fhAPI.PutRecord(&firehose.PutRecordInput{
-			DeliveryStreamName: aws.String(al.fhStream),
-			Record:             &firehose.Record{Data: bs},
-		})
-		if err != nil {
-			return err
-		}
-		return nil
-	}); err != nil {
-		return 0, fmt.Errorf("PutRecords: %v", err)
+	al.batchBytes += len(bs)
+	al.batch = append(al.batch, &firehose.Record{Data: bs})
+	if len(al.batch) == al.maxBatchRecords || al.batchBytes > int(0.9*float64(al.maxBatchBytes)) {
+		return len(bs), al.sendBatch()
 	}
 	return len(bs), nil
 }
 
+// Close flushes all logs to Firehose.
+func (al *Logger) Close() error {
+	al.mu.Lock()
+	defer al.mu.Unlock()
+	if len(al.batch) > 0 {
+		return al.sendBatch()
+	}
+	return nil
+}
+
+func (al *Logger) sendBatch() error {
+	defer func() {
+		// reset the batch state no matter what, since
+		// errors will only get worse if we keep adding to
+		// the batch (i.e. we'll exceed record or byte limits)
+		al.batch = []*firehose.Record{}
```

Comment:
This is the only part that we're holding the lock for. But it means that while `sendBatch` is running, new requests will get blocked waiting for the lock. Can we instead do something like this in `Close` (and similar in `Write`)?

```
var batch []*firehose.Record
al.mu.Lock()
if len(al.batch) > 0 {
	batch = al.batch
	al.batch = []*firehose.Record{}
	al.batchBytes = 0
}
al.mu.Unlock()
return sendBatch(batch)
```

Reply:
Discussed yesterday with Raf. There is potential concern about how it could be easy to accidentally send the shared batch, rather than copies.

I think doing it this way is sufficiently worthwhile from the perspective of not blocking, but I added a comment, and made `sendBatch` not a method to prevent it being super easy to accidentally do stuff with `al.batch` inside the function.